### PR TITLE
Plane: fixed terrain RTL rally bug

### LINKED ISSUE #25157: Plane: RTL with rally point does not track terrain
when using RTL with a rally point with high points in the terrain between current location and the rally point, with TERRAIN_FOLLOW=1 the plane sometimes does not climb over the terrain as it should. As yet it is unclear under what conditions it fails.

Reproduced in 4.4.1 stable in SITL with the following log:
http://uav.tridgell.net/tmp/00000013.BIN

### PR DESCRIPTION
**Pull request by tridge**
this fixes a bug where if the terrain database cache does not have the tile for the location of a rally point then RTL to the rally point with TERRAIN_FOLLOW=1 will not track terrain
    
The underlying issue is that Location::loc.change_alt_frame() will return false if the location is not in the terrain memory cache. We can't just extrapolate as the rally point could be in a totally different terrain area to the current location. So instead we set it as terrain_following_pending and fix it as soon as the terrain cache is filled.

this fixes #25157

### CODE REVIEW

File: Tools/autotest/common.py [hunk 1]
```diff
@@ -7084,6 +7084,35 @@ def assert_rc_channel_value(self, channel, value):
...
+                                      0, # target component
+                                      seq, # sequence number
+                                      total, # total count
+                                      int(loc.lat * 1e7),
+                                      int(loc.lng * 1e7),
+                                      loc.alt, # relative alt
+                                      0, # "break" alt?!
+                                      0, # "land dir"
+                                      0) # flags
+
```
Comment (peterbarker) [inline Tools/autotest/common.py:7115]:
```suggestion
```

File: Tools/autotest/arduplane.py [hunk 2]
```diff
@@ -1836,6 +1836,85 @@ def FenceRetRally(self):
...
+
+        guided_loc = mavutil.location(-35.39723762, 149.07284612, 99.0, 0)
+        rally_loc = mavutil.location(-35.3654952000, 149.1558698000, 100, 0)
+
+        terrain_wait_path(self.homeloc, rally_loc, 10)
+
+        # set a rally point to the west of home
+        self.set_parameter("RALLY_TOTAL", 1)
+        self.add_rally_point(rally_loc, 0, 1)
+        self.delay_sim_time(1)
```
Comment (peterbarker) [inline Tools/autotest/arduplane.py:1884]:
```suggestion
        self.upload_rally_points_from_locations([rally_loc])
```

... this method mimics 2 other methods we have for waypoints and fence items.

File: Tools/autotest/arduplane.py [hunk 3]
```diff
@@ -1836,6 +1836,85 @@ def FenceRetRally(self):
...
+        self.progress("Done takeoff")
+
+        self.install_message_hook_context(terrain_following_above_80m)
+
+        self.change_mode("GUIDED")
+        self.do_reposition(guided_loc, frame=mavutil.mavlink.MAV_FRAME_GLOBAL_TERRAIN_ALT)
+        self.progress("Flying to guided location")
+        self.wait_location(guided_loc,
+                           accuracy=200,
+                           target_altitude=None,
```
Comment (peterbarker) [inline Tools/autotest/arduplane.py:1901]:
```suggestion
```

File: Tools/autotest/arduplane.py [hunk 4]
```diff
@@ -1836,6 +1836,85 @@ def FenceRetRally(self):
...
+                           timeout=600)
+
+        self.progress("Reached guided location")
+        self.set_parameter("RALLY_LIMIT_KM", 50)
+        self.change_mode("RTL")
+        self.progress("Flying to rally point")
+        self.wait_location(rally_loc,
+                           accuracy=200,
+                           target_altitude=None,
```
Comment (peterbarker) [inline Tools/autotest/arduplane.py:1910]:
```suggestion
```

File: Tools/autotest/arduplane.py [hunk 5]
```diff
@@ -1836,6 +1836,85 @@ def FenceRetRally(self):
...
+            while True:
+                now = self.get_sim_time_cached()
+                if now - tstart > 60:
+                    raise NotAchievedException("Did not get correct required terrain")
+                for i in range(steps):
+                    lat = loc1.lat + i * (loc2.lat-loc1.lat)/steps
+                    lon = loc1.lng + i * (loc2.lng-loc1.lng)/steps
+                    self.mav.mav.terrain_check_send(int(lat*1.0e7), int(lon*1.0e7))
+
+                report = self.mav.recv_match(type='TERRAIN_REPORT', blocking=True, timeout=60)
```
Comment (peterbarker) [inline Tools/autotest/arduplane.py:1867]:
```suggestion
                report = self.assert_receive_message('TERRAIN_REPORT', timeout=60)
```

File: Tools/autotest/arduplane.py [hunk 6]
```diff
@@ -1836,6 +1836,85 @@ def FenceRetRally(self):
         self.do_fence_disable() # Disable fence so we can land
         self.fly_home_land_and_disarm() # Pack it up, we're going home.
 
+    def TerrainRally(self):
+        """ Tests terrain follow with a rally point """
+        self.context_push()
+        self.install_terrain_handlers_context()
+        self.progress("Testing FENCE_ACTION_RTL with fence rally point")
```
Comment (peterbarker) [inline Tools/autotest/arduplane.py:1843]:
```suggestion
```

**Comment by peterbarker** [comment]
... should have mentioned https://github.com/ArduPilot/ardupilot/pull/25168 contains that new method.

File: Tools/autotest/common.py (hunk 1 above)
Comment (tridge) [reply Tools/autotest/common.py:7115]: ignored, no idea what you're suggesting or why

File: Tools/autotest/common.py [hunk 7]
```diff
@@ -7084,6 +7084,35 @@ def assert_rc_channel_value(self, channel, value):
             raise NotAchievedException("Expected %s to be %u got %u" %
                                        (channel, value, m_value))
 
+    def do_reposition(self,
+                      loc,
+                      frame=mavutil.mavlink.MAV_FRAME_GLOBAL_RELATIVE_ALT_INT):
```
[the remark this answers is not in the document]
Comment (tridge) [reply Tools/autotest/common.py:7089]: changed

File: Tools/autotest/arduplane.py (hunk 4 above)
Comment (tridge) [reply Tools/autotest/arduplane.py:1910]: ignored. I want to make it obvious that we don't care about altitude, it isn't obvious from the function name

File: Tools/autotest/arduplane.py (hunk 3 above)
Comment (tridge) [reply Tools/autotest/arduplane.py:1901]: ignored, I want it obvious in caller we don't care about alt

File: Tools/autotest/arduplane.py (hunk 4 above)
Comment (peterbarker) [reply Tools/autotest/arduplane.py:1910]: Yes, we should probably rename this function to `wait_lat_lng` and create a new `wait_location` which honours the height in `loc`

File: Tools/autotest/arduplane.py (hunk 5 above)
Comment (tridge) [reply Tools/autotest/arduplane.py:1867]: done

File: Tools/autotest/arduplane.py (hunk 6 above)
Comment (tridge) [reply Tools/autotest/arduplane.py:1843]: done

File: Tools/autotest/arduplane.py (hunk 2 above)
Comment (tridge) [reply Tools/autotest/arduplane.py:1884]: done

File: Tools/autotest/arduplane.py [hunk 8]
```diff
@@ -1836,6 +1836,85 @@ def FenceRetRally(self):
...
+        self.add_rally_point(rally_loc, 0, 1)
+        self.delay_sim_time(1)
+
+        self.set_parameter("TKOFF_ALT", 100)
+        self.change_mode("TAKEOFF")
+        self.wait_ready_to_arm()
+        self.arm_vehicle()
+        self.set_parameter("TERRAIN_FOLLOW", 1)
+        self.wait_altitude(90, 120, timeout=30, relative=True)
+        self.progress("Done takeoff")
```
[the remark this answers is not in the document]
Comment (tridge) [reply Tools/autotest/arduplane.py:1892]: ignored. I want a very specific sequence to mimic a specific log where I reproduced the problem

File: Tools/autotest/common.py (hunk 1 above)
Comment (peterbarker) [reply Tools/autotest/common.py:7116]: The method is unused!

All the red stuff means "remove these lines".

**Review by peterbarker — APPROVED** [review]
LGTM

File: ArduPlane/Plane.h [hunk 9]
```diff
@@ -735,6 +735,9 @@ class Plane : public AP_Vehicle {
         // are we trying to follow terrain?
         bool terrain_following;
 
+        // are we waiting to load terrain data to init terrain following
+        bool terrain_following_pending;
```
Comment (nexton-winjeel) [inline ArduPlane/Plane.h:739]: @tridge: Is there any issue with this getting out of sync with `terrain_following`? Might be better to combine them into an `enum`?

Comment (tridge) [reply ArduPlane/Plane.h:739]: i'd be happy to change that in a future PR, yes

**Pull request merged**